### PR TITLE
Feature/window_generator

### LINKED ISSUE #34: Window Frames Generator from File
**Is your feature request related to a problem? Please describe.**
In order to work with large trajectories that do not fit into the RAM, as window generator function directly from a trajectory would be really helpful.

**Describe the solution you'd like**
Add new method to `TrajectoryReader`, which calls the  `frame_generator` method.

### PR DESCRIPTION
**Pull request by galjos**
This feature closes #34

### CODE REVIEW

File: PQAnalysis/io/traj_file/trajectory_reader.py [hunk 1]
```diff
@@ -107,57 +120,271 @@ def read(self, topology: Topology | None = None) -> Trajectory:
...
+            of the trajectory.
+
         Yields
         ------
         Generator[AtomicSystem]
             The frames of the trajectory.
         """
+
+        if trajectory_stop is None:
+            trajectory_stop = self.length_of_traj
```
Comment (97gamjak) [inline PQAnalysis/io/traj_file/trajectory_reader.py:164]: see comment in __init__

File: PQAnalysis/io/traj_file/trajectory_reader.py [hunk 2]
```diff
@@ -107,57 +120,271 @@ def read(self, topology: Topology | None = None) -> Trajectory:
...
+
+        # If trajectory_start is greater than or equal to trajectory_stop,
+        # raise an IndexError
+        if trajectory_start >= trajectory_stop:
+            self.logger.error(
+                "start index is greater than or equal to the stop index",
+                exception=IndexError,
+            )
+
+        # Track the number of frames that have been yielded
```
Comment (97gamjak) [inline PQAnalysis/io/traj_file/trajectory_reader.py:190]: isn't this the number of frames that have been read and not that have been yielded?

File: PQAnalysis/io/traj_file/trajectory_reader.py [hunk 3]
```diff
@@ -107,57 +120,271 @@ def read(self, topology: Topology | None = None) -> Trajectory:
...
+            set to the length of the trajectory.
+
+        Raises
+        ------
+        IndexError
+            If window_size is less than 1 or greater than the length of
+            the trajectory.
+            If window_gap is less than 1 or greater than the length of
+            the trajectory.
+            If window_size is greater than trajectory_stop - trajectory_start.
```
Comment (97gamjak) [inline PQAnalysis/io/traj_file/trajectory_reader.py:293]: missing raises

File: PQAnalysis/io/traj_file/trajectory_reader.py [hunk 4]
```diff
@@ -107,57 +120,271 @@ def read(self, topology: Topology | None = None) -> Trajectory:
...
+
+        Warning
+        -------
+        If not all frames are included in the windows, a warning is issued.
+
+        Yields
+        ------
+        Generator[Trajectory]
+            An generator over the windows of the trajectory with the specified
```
Comment (97gamjak) [inline PQAnalysis/io/traj_file/trajectory_reader.py:302]:
```suggestion
            A generator over the windows of the trajectory with the specified
```

File: PQAnalysis/io/traj_file/trajectory_reader.py [hunk 5]
```diff
@@ -107,57 +120,271 @@ def read(self, topology: Topology | None = None) -> Trajectory:
...
+        # Check if all frames are included in the windows
+        # Length of the trajectory - window_size should be divisible by window_gap
+        if ((trajectory_stop - trajectory_start) - window_size) % window_gap != 0:
+            self.logger.warning(
+                "Not all frames are included in the windows. Check the window size and gap."
+            )
+
+        generator = self.frame_generator(
+            trajectory_start=trajectory_start, trajectory_stop=trajectory_stop
+        )
```
Comment (97gamjak) [inline PQAnalysis/io/traj_file/trajectory_reader.py:366]:
```suggestion
        generator = self.frame_generator(
            trajectory_start=trajectory_start,
            trajectory_stop=trajectory_stop
        )
```

File: PQAnalysis/io/traj_file/trajectory_reader.py [hunk 6]
```diff
@@ -107,57 +120,271 @@ def read(self, topology: Topology | None = None) -> Trajectory:
...
+        # reads first window and converts it to a queue
+        window = Trajectory([next(generator) for _ in range(window_size)])
+
+        # yield the first window
+        yield window.copy()
+
+        # generate the rest of the windows up to trajectory_stop
+        for _ in range(
+            trajectory_start + window_gap, trajectory_stop - window_size + 1, window_gap
+        ):
```
Comment (97gamjak) [inline PQAnalysis/io/traj_file/trajectory_reader.py:377]:
```suggestion
        for _ in range(
            trajectory_start + window_gap,
            trajectory_stop - window_size + 1,
            window_gap
        ):
```

File: PQAnalysis/io/traj_file/trajectory_reader.py [hunk 7]
```diff
@@ -107,57 +120,271 @@ def read(self, topology: Topology | None = None) -> Trajectory:
...
+            self.logger.warning(
+                "Not all frames are included in the windows. Check the window size and gap."
+            )
+
+        generator = self.frame_generator(
+            trajectory_start=trajectory_start, trajectory_stop=trajectory_stop
+        )
+
+        # reads first window and converts it to a queue
+        window = Trajectory([next(generator) for _ in range(window_size)])
```
Comment (97gamjak) [inline PQAnalysis/io/traj_file/trajectory_reader.py:369]: pylint gives me here a [problem `stop-iteration-return / R1708`](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/stop-iteration-return.html).

So now here all edge cases are covered - should we disable here the pylint problem or would it be best to add further checks for future sanity?

Comment (galjos) [reply PQAnalysis/io/traj_file/trajectory_reader.py:369]: I can't quite figure out, how to build in a StopIteration raise. The frame_generator automatically stops when the trajectory is finished. In my opinion pylint should be disabled for this special case.

File: PQAnalysis/io/traj_file/trajectory_reader.py (hunk 2 above)
Comment (galjos) [reply PQAnalysis/io/traj_file/trajectory_reader.py:190]: True. It was different in the initial implementation.

File: PQAnalysis/io/traj_file/trajectory_reader.py [hunk 8]
```diff
@@ -66,6 +76,9 @@ def __init__(self,
         self.md_format = MDEngineFormat(md_format)
         self.frame_reader = FrameReader(md_format=self.md_format)
 
+        # Calculate the number of frames in the trajectory
+        self.length_of_traj = self.calculate_number_of_frames()
```
[the remark this answers is not in the document]
Comment (galjos) [reply PQAnalysis/io/traj_file/trajectory_reader.py:80]: True. Should the generator read the frames that are provided at the specific point in time? So, when the generator is called this specific trajectory should be read?

Suggestion:
Maybe PQ should write a 'end-of-simulation' flag to the files, then all file readers could wait till they know that the simulation is finished.

[the remark this answers is not in the document]
Comment (97gamjak) [reply PQAnalysis/io/traj_file/trajectory_reader.py]: > True. Should the generator read the frames that are provided at the specific point in time? So, when the generator is called this specific trajectory should be read?
> 
> Suggestion: Maybe PQ should write a 'end-of-simulation' flag to the files, then all file readers could wait till they know that the simulation is finished.

Could be a good solution, but then where should this flag be written? At the end of the last line as an additional comment

[the remark this answers is not in the document]
Comment (galjos) [reply PQAnalysis/io/traj_file/trajectory_reader.py]: Yes, maybe.

**Pull request merged**